Add Unused Parameter Warnings

n01r opened the pull request:
Adds warnings about unused parameters like in WarpX as requested in #120 

- [x] Can you please check if `amrex.abort_on_unused_inputs = 1` works already as well?
- [x] This one needs to be documented in both WarpX and AMReX docs, please.
- [x] We need to please document `warpx.abort_on_warning_threshold` (or equivalent) in ImpactX.
- [x] implemented it like in WarpX but it does not abort, yet, when I do some warning logger testing as described in the WarpX docs
- [x] move impl to separate `.cpp` file
- [x] add Python control
- [x] add docs for Python control
- ~~[ ] add Python tests for unused inputs and for warning in `tests/python/test_impactx.py`~~ skipped for follow-up since more testing is needed

n01r committed:
Add Unused Parameter Warnings

ax3l commented:
Thanks!

- Can you please check if `amrex.abort_on_unused_inputs = 1` works already as well?
- This one needs to be documented in both WarpX and AMReX docs, please.
- We need to please document `warpx.abort_on_warning_threshold` (or equivalent) in ImpactX.

ax3l commented on src/ImpactX.cpp:
```diff
@@ -194,6 +198,17 @@ namespace impactx
+                    amrex::ParmParse().QueryUnusedInputs();
+
+                    //Print the warning list right after the first step.
+                    amrex::Print() <<
+                                   ablastr::warn_manager::GetWMInstance().PrintGlobalWarnings("FIRST STEP");
```
Might be better:
```suggestion
                    amrex::Print() << ablastr::warn_manager::GetWMInstance()
                                      .PrintGlobalWarnings("FIRST STEP");
```

As in WarpX, you can also print again after the last step.

n01r committed:
update formatting

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

n01r commented:
~~Setting `amrex.abort_on_unused_inputs = 1` and having an unused input does not abort the sim, yet.~~

It works, but it aborts *after* the sim runs through which first confused me.

ax3l commented:
Thanks, please don't forget adding user-facing docs.

n01r committed:
add user-facing docs

n01r commented:
> We need to please document warpx.abort_on_warning_threshold (or equivalent) in ImpactX.

I think this additional implementation of these options and maybe it justifies creating a function `ImpactX::ReadParameters` like the existing one in WarpX.

ax3l commented:
The nice thing about `amrex::ParmParse` is, that you can read parameters anywhere. For which current collection of parameters do you suggest grouping in a `ReadParameters`?

Please let me know when these items are ready (https://github.com/ECP-WarpX/impactx/pull/223#issuecomment-1232275511) and this can be merged :) I moved them as to-do's in the PR description.

n01r committed:
Add abort_on_warning_threshold

and add user-facing docs

n01r commented:
> The nice thing about amrex::ParmParse is, that you can read parameters anywhere. For which current collection of parameters do you suggest grouping in a ReadParameters?

Why not sort all? In WarpX it seems they are all grouped into a `ReadParameters` function and every group has its own context like:
```cpp
    {
        ParmParse pp_<context>("< context >");
        pp_<context>.query("name-of-some-option")
    }
```

> Please let me know when these items are ready (https://github.com/ECP-WarpX/impactx/pull/223#issuecomment-1232275511) and this can be merged :) I moved them as to-do's in the PR description.
> - Can you please check if amrex.abort_on_unused_inputs = 1 works already as well?

this works already and user-facing docs have been added :)

> - This one needs to be documented in both WarpX and AMReX docs, please.

- user-facing docs for WarpX: https://github.com/ECP-WarpX/WarpX/pull/3364
- ~ for AMReX: https://github.com/AMReX-Codes/amrex/pull/2937

> - We need to please document warpx.abort_on_warning_threshold (or equivalent) in ImpactX.

- implemented it like in WarpX but it does not abort, yet, when I do some warning logger testing as described in the WarpX docs

ax3l commented on src/ImpactX.cpp:
```diff
@@ -194,6 +227,17 @@ namespace impactx
                 }
 
+                // inputs: unused parameters (e.g. typos) check after step 1 has finished
+                if (!early_params_checked) {
```
lets put this in a member function: `early_param_check ()`

ax3l commented on src/ImpactX.cpp:
```diff
@@ -59,6 +62,8 @@ namespace impactx
         //   before we start the evolve loop, we are in "step 0" (initial state)
         int global_step = 0;
 
+        bool early_params_checked = false; // check typos in inputs after step 1
```
Reset the variable here, make it otherwise a member variable.
pls move comment as its docstring
```suggestion
        early_params_checked = false; // check typos in inputs after step 1
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -108,6 +113,34 @@ namespace impactx
         pp_algo.queryAdd("space_charge", space_charge);
         amrex::Print() << " Space Charge effects: " << space_charge << "\n";
 
+        amrex::ParmParse pp_impactx("impactx");
```
This should not be in `evolve ()` since that is too late.
It could be a function called at the end of `ImpactX::ImpactX`

n01r committed:
Refactor warning code into member functions

ax3l commented on src/ImpactX.cpp:
```diff
@@ -33,6 +36,41 @@ namespace impactx
+
+    }
+
+    void ImpactX::warn_logger_control()
```
To improve compile time, can you put this in a separate file please, e.g., `src/initialization/Warnings.cpp`? :)

ax3l commented on src/ImpactX.cpp:
```diff
@@ -51,6 +89,21 @@ namespace impactx
         amrex::UtilCreateCleanDirectory("diags", true);
     }
 
+    void ImpactX::early_param_check()
```
this can go in the same file noted above :)

ax3l committed:
Fix: Test Warnings After Setting Threshold

ax3l committed:
Examples/Tests: Abort if Warnings Encountered

n01r committed:
Extract warning logger control to separate file

- create initialization/Warnings.cpp
- put early_param_check() and warning_logger_control() there

n01r committed:
Expose warning logger parameters to pybind

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented:
> In WarpX it seems they are all grouped into a ReadParameters function ...

Minutes for what we discussed: We need no static member variables for states because we store our options directly in `amrex::ParmParse`. This also keeps us in sync which options take precedence for control from the Python side. Querying `ParmParse` is fast and can be done in-place instead of at the init stage only.

ax3l commented on src/ImpactX.H:
```diff
@@ -86,10 +86,20 @@ namespace impactx
+         *
+         * Input variables are: ``always_warn_immediately`` and ``abort_on_warning_threshold``.
+         */
+        void warn_logger_control ();
```
Maybe this is a bit more descriptive:
```suggestion
        void init_warning_logger ();
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -36,6 +36,10 @@ namespace impactx
         //       amr.n_cells from user inputs
+
+        // query input for warning logger variables and set up warning logger accordingly
+        warn_logger_control();
```
```suggestion
        init_warning_logger();
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -64,6 +68,23 @@ namespace impactx
+        // check typos in inputs after step 1
+        bool early_params_checked = false;
+
+        // count particles - if no particles are found in our particle container, then a lot of
```
Please remove this check - this should now be part of `validate()` already, isn't it? Otherwise please move there.

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+#include <AMReX_BLProfiler.H>
+
+
+void impactx::ImpactX::warn_logger_control ()
```
```suggestion
void impactx::ImpactX::init_warning_logger ()
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+
+void impactx::ImpactX::warn_logger_control ()
+{
+    amrex::ParmParse pp_impactx("impactx");
```
Please add include for `amrex::ParmParse`

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+    ablastr::warn_manager::GetWMInstance().debug_read_warnings_from_input(pp_impactx);
+}
+
+void impactx::ImpactX::early_param_check()
```
AMReX style convention:
```suggestion
void impactx::ImpactX::early_param_check ()
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+    amrex::Print() << "\n";
+    amrex::ParmParse().QueryUnusedInputs();
+
+    //Print the warning list right after the first step.
```
```suggestion
    // Print the warning list right after the first step.
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+            amrex::Abort(ablastr::utils::TextMsg::Err(str_abort_on_warning_threshold
+                                                      +"is not a valid option for impactx.abort_on_warning_threshold (use: low, medium or high)"));
+        }
+        amrex::AllPrint() << "str_abort_on_warning_threshold=" << str_abort_on_warning_threshold << "\n";
+        ablastr::warn_manager::GetWMInstance().SetAbortThreshold(abort_on_warning_threshold);
```
Please break overlong lines for readability, remove the debug print statement:
```suggestion
        ablastr::warn_manager::GetWMInstance()
            .SetAbortThreshold(abort_on_warning_threshold);
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+{
+    amrex::ParmParse pp_impactx("impactx");
+
+    // Set the flag to control if ImpactX has to emit a warning message as soon as a warning is recorded
```
readability:
```suggestion
    // Set the flag to control if ImpactX has to emit a warning message as
    // soon as a warning is recorded
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+    pp_impactx.query("always_warn_immediately", always_warn_immediately);
+    ablastr::warn_manager::GetWMInstance().SetAlwaysWarnImmediately(always_warn_immediately);
+
+    // Set the WarnPriority threshold to decide if ImpactX has to abort when a warning is recorded
```
Readability:
```suggestion
    // Set the WarnPriority threshold to decide if ImpactX has to abort
    // when a warning is recorded
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+    ablastr::warn_manager::GetWMInstance().SetAlwaysWarnImmediately(always_warn_immediately);
+
+    // Set the WarnPriority threshold to decide if ImpactX has to abort when a warning is recorded
+    if(std::string str_abort_on_warning_threshold = "";
```
Please include `<string>`

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+    // Set the WarnPriority threshold to decide if ImpactX has to abort when a warning is recorded
+    if(std::string str_abort_on_warning_threshold = "";
+            pp_impactx.query("abort_on_warning_threshold", str_abort_on_warning_threshold)){
+        std::optional<ablastr::warn_manager::WarnPriority> abort_on_warning_threshold = std::nullopt;
```
Please include `<optional>`

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,61 @@
+        else if (str_abort_on_warning_threshold == "low")
+            abort_on_warning_threshold = ablastr::warn_manager::WarnPriority::low;
+        else {
+            amrex::Abort(ablastr::utils::TextMsg::Err(str_abort_on_warning_threshold
```
Please throw a `std::runtime` error instead of calling `amrex::Abort` here.
We want to be able to catch this in all cases.

ax3l commented on src/python/ImpactX.cpp:
```diff
@@ -181,6 +181,35 @@ void init_ImpactX(py::module& m)
              "The minimum number of digits (default: 6) used for the step\n"
              "number appended to the diagnostic file names."
          )
+        .def("set_abort_on_warning_threshold",
```
Please add tests for this functionality as noted in the PR description

n01r replied:
**update:** skipping Python tests for this PR as discussed this morning and doing them in a follow-up

ax3l commented on src/python/ImpactX.cpp:
```diff
@@ -181,6 +181,35 @@ void init_ImpactX(py::module& m)
+                 " as soon as it is generated."
+        )
+        // TODO this is an integer with 0 or 1 - can I just make this a boolean here?
+        .def("set_abort_on_unused_parameters",
```
For all three: as discussed, please use `def_property` as in `dynamic_size`

n01r committed:
Implement reviewer changes and python control

n01r committed:
Add docs for Python interface

add docs for:
  - abort_on_warning_threshold
  - abort_on_unused_inputs
  - always_warn_immediately

ax3l commented on src/ImpactX.cpp:
```diff
@@ -36,6 +38,14 @@ namespace impactx
+        // inputfile for debugging & testing purposes.
+        amrex::ParmParse pp_impactx("impactx");
+        ablastr::warn_manager::GetWMInstance().
+                debug_read_warnings_from_input(pp_impactx);
```
I think we can actually skip this here - this is a self-test for the warning logger and can be done upstream in WarpX/ABLASTR.
```suggestion
```

ax3l commented on src/ImpactX.H:
```diff
@@ -86,10 +86,20 @@ namespace impactx
 
+        /** Check inputs for unused parameters (e.g. typos) after step 1
+         */
+        void early_param_check ();
```
Little idea to make the call-site more compact:
```suggestion
        /** Check inputs for unused parameters (e.g. typos) after step 1
         *
         * @return true if successful
         */
        void early_param_check ();
```

ax3l commented on src/ImpactX.cpp:
```diff
@@ -199,6 +212,9 @@ namespace impactx
                 }
 
+                // inputs: unused parameters (e.g. typos) check after step 1 has finished
+                if (!early_params_checked) { early_param_check(); early_params_checked = true;}
```
See API change:
```suggestion
                if (!early_params_checked) { early_params_checked = early_param_check(); }
```

(we could have also passed by reference, but I like being explicit a bit more.)

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,64 @@
+    }
+}
+
+void impactx::ImpactX::early_param_check ()
```
```suggestion
bool ImpactX::early_param_check ()
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,64 @@
+    // Print the warning list right after the first step.
+    amrex::Print() << ablastr::warn_manager::GetWMInstance()
+            .PrintGlobalWarnings("FIRST STEP");
+
```
```suggestion

    return true;
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,64 @@
+    amrex::Print() << ablastr::warn_manager::GetWMInstance()
+            .PrintGlobalWarnings("FIRST STEP");
+
+}
```
```suggestion
}
} // namespace impactx
```

ax3l commented on src/initialization/Warnings.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <string>
+
+
+void impactx::ImpactX::init_warning_logger ()
```
Minimally more compact:
```suggestion
namespace impactx
{
void ImpactX::init_warning_logger ()
```

n01r committed:
Apply suggestions from code review

Thanks for the suggestions!

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

n01r commented on src/ImpactX.H:
```diff
@@ -86,10 +86,22 @@ namespace impactx
+         *
+         * @return true if successful
+         */
+        void early_param_check ();
```
```suggestion
        bool early_param_check ();
```

n01r committed:
fix signature
